### PR TITLE
Import books from Goodreads

### LINKED ISSUE #77: Import books from Goodreads
**Describe the solution you'd like**
Take in a csv file that was exported from Goodreads and import it into the app.

If a particular book was in a bookshelf that corresponds to a 'to read', reading or read predefined book shelf, the import tool should be able to automatically move the book into that shelf. Otherwise, the user needs to be prompted and asked which shelf to put the book in.

An imported book should be bound by the relevant fields.

**Additional context**
[Sample file](https://www.goodreads.com/assets/sample_export.csv)

- Refer to Vaadin's documentation on the [upload component](https://vaadin.com/components/vaadin-upload/java-examples)

- We can ignore the average rating, year published and original publication year columns
- If a date read has been speciifed, the book should go in the 'read' shelf
- Until the publisher and binding (we call it 'book type') attributes are added to the `Book` entity, we can ignore it 
- If a book does not have a title and/or an author, we do not add it
- Since we're using a 10-point scale, we should scale any ratings by 2 (Goodreads uses a 5-point scale with no half stars)
- If a user has a book shelf that we don't have for the user, we need to add it
- If there are any errors in parsing the file (e.g. a missing required field), ignore that row, but try to add as many books as possible (not atomic)

In the future, we will provide the user with an option to review the imports. This would allow a user to make any changes (e.g. update their rating)

### PR DESCRIPTION
**Pull request by labradav**
## Summary of change

Adds an import button in SettingsView.java to upload Goodreads books using a csv file.

## Additional context

Screenshot of the final Settings view:

Screenshot 2020-11-01 at 7 22 53 PM

## Related issue

Closes #77 

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request
  - [x] Included a screenshot(s) if a UI change was involved (it may look different on a reviewer's device)

- [x] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Resolved any merge conflicts

For any of the optional checkboxes (e.g. the screenshots one), still check it if it does not apply.

If in doubt, get in touch with us via our [Slack workspace](https://join.slack.com/t/teambookproject/shared_invite/zt-i7lept44-rgJ9yB0A2vedJTLyyfkjKQ)

### CODE REVIEW

**Comment by knjk04** [comment]
@labradav How are you getting on with this?

**Comment by labradav** [comment]
@knjk04 I have added the upload button in the Settings page which will be used to upload the file. The file is getting uploaded successfully. Since you suggested that we can periodically review the PR, I was hoping if you have any suggestions for the current changes. Meanwhile, I'm working on converting the contents of the uploaded file to the Book entity and adding them to the to read shelf by default. Also, I will be using `jackson-dataformat` to parse the file to a DTO projection which can then be mapped to the Book entity.

**Comment by knjk04** [comment]
@labradav Sure, I'll take a look

File: src/main/java/com/karankumar/bookproject/backend/utils/CsvUtils.java
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/utils/CsvUtils.java:25]: Could you please add a test for this?

File: src/main/java/com/karankumar/bookproject/backend/utils/CsvUtils.java
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/utils/CsvUtils.java:28]: What happens if there is any IO exception thrown, I guess we should handle the exception and not throw back?

File: src/main/java/com/karankumar/bookproject/backend/utils/CsvUtils.java
Comment (amit1307) [inline src/main/java/com/karankumar/bookproject/backend/utils/CsvUtils.java:25]: What would happen if the file type is not csv? Do we want to fail fast with some error message about the file type?

**Review by amit1307 — CHANGES_REQUESTED** [review]
Sorry but to me it seems like the PR is not complete, when I look at the corresponding issue it says:

- Take in a csv file that was exported from Goodreads and import it into the app.
But in this PR I don't see anywhere the imported data is saved in the DB.

- Moreover the issue describes lots of edge cases that we need to take care of while importing like 
`If a particular book was in a bookshelf that corresponds to a to read, reading or read predefined book shelf, the import tool should be able to automatically move the book into that shelf. Otherwise, the user needs to be prompted and asked which shelf to put the book in.`

- There are a few other cases mentioned in the issue description that we need to take care of but I am not able to see any code changes for them.

Please let me know if I am missing anything.

**Comment by labradav** [comment]
> Sorry but to me it seems like the PR is not complete, when I look at the corresponding issue it says:
> 
> * Take in a csv file that was exported from Goodreads and import it into the app.
>   But in this PR I don't see anywhere the imported data is saved in the DB.
> * Moreover the issue describes lots of edge cases that we need to take care of while importing like
>   `If a particular book was in a bookshelf that corresponds to a to read, reading or read predefined book shelf, the import tool should be able to automatically move the book into that shelf. Otherwise, the user needs to be prompted and asked which shelf to put the book in.`
> * There are a few other cases mentioned in the issue description that we need to take care of but I am not able to see any code changes for them.
> 
> Please let me know if I am missing anything.

Yes @amit1307 , the changes are not complete yet. The idea was to periodically review it rather than doing it all at once which was suggested by @knjk04 . I'm working on the changes which will actually validate and persist the contents of the file in the db.

**Comment by amit1307** [comment]
> > Sorry but to me it seems like the PR is not complete, when I look at the corresponding issue it says:
> > 
> > * Take in a csv file that was exported from Goodreads and import it into the app.
> >   But in this PR I don't see anywhere the imported data is saved in the DB.
> > * Moreover the issue describes lots of edge cases that we need to take care of while importing like
> >   `If a particular book was in a bookshelf that corresponds to a to read, reading or read predefined book shelf, the import tool should be able to automatically move the book into that shelf. Otherwise, the user needs to be prompted and asked which shelf to put the book in.`
> > * There are a few other cases mentioned in the issue description that we need to take care of but I am not able to see any code changes for them.
> > 
> > Please let me know if I am missing anything.
> 
> Yes @amit1307 , the changes are not complete yet. The idea was to periodically review it rather than doing it all at once which was suggested by @knjk04 . I'm working on the changes which will actually validate and persist the contents of the file in the db.

ah okay, cool..sorry I missed that bit, should have checked the earlier comment on the PR.

File: src/main/java/com/karankumar/bookproject/backend/utils/CsvUtils.java
Comment (labradav) [reply src/main/java/com/karankumar/bookproject/backend/utils/CsvUtils.java:25]: I'm not entirely sure if we can add a validation here for file type, as we are using InputStream here and the underlying implementation may not necessarily be a FileInputStream.

File: src/main/java/com/karankumar/bookproject/backend/utils/CsvUtils.java
Comment (labradav) [reply src/main/java/com/karankumar/bookproject/backend/utils/CsvUtils.java:28]: If we do handle the exception, and return for instance an empty list, the user will be oblivious to the fact that whether the import was unsuccessful or the file was indeed empty. That is why I have thrown it so that the client can handle it according to the appropriate use case.

File: src/main/java/com/karankumar/bookproject/backend/utils/CsvUtils.java
Comment (labradav) [reply src/main/java/com/karankumar/bookproject/backend/utils/CsvUtils.java:25]: Sure.

**Comment by knjk04** [comment]
@labradav How are you getting on with this?

**Comment by labradav** [comment]
> @labradav How are you getting on with this?

Hey @knjk04 , I have been a bit occupied with work and the holidays as well in the past week. I will try to finish all the requirements by the end of the week. My apologies for the delay.

**Comment by knjk04** [comment]
@labradav That's fine, thanks for the update

File: src/main/java/com/karankumar/bookproject/backend/dto/GoodreadsBookImport.java
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/backend/dto/GoodreadsBookImport.java:1]: Could you please attach our [copyright notice](https://github.com/Project-Books/book-project/blob/master/COPYRIGHT) file to the top of this file?

File: src/main/java/com/karankumar/bookproject/backend/service/ImportService.java
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/backend/service/ImportService.java:1]: copyright notice here

File: src/main/java/com/karankumar/bookproject/backend/service/ImportService.java
Comment (knjk04) [inline src/main/java/com/karankumar/bookproject/backend/service/ImportService.java:123]: As the scale factor is always 2, could we make this a constant, instead?

File: src/test/java/com/karankumar/bookproject/backend/utils/CsvUtilsTest.java
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/utils/CsvUtilsTest.java:1]: copyright notice

File: src/test/java/com/karankumar/bookproject/backend/utils/CsvUtilsTest.java
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/utils/CsvUtilsTest.java:20]: Could you change this to use AssertJ and add given/when/then comments?

**Review by knjk04 — CHANGES_REQUESTED** [review]
I've left some preliminary feedback from a quick look. Also, I noticed the build is failing due to failing tests? Could you take a look into this? Let us know over Slack if you need a hand

File: src/main/java/com/karankumar/bookproject/backend/service/ImportService.java
Comment (labradav) [reply src/main/java/com/karankumar/bookproject/backend/service/ImportService.java:123]: That is true for goodreads. But the service and inturn the method is source agnostic, that is why I have taken a parameter here as the scale factor might be different for various third party data stores.

Comment (knjk04) [reply src/main/java/com/karankumar/bookproject/backend/service/ImportService.java:123]: That makes sense. Could you create a variable and pass it into the method to make it a little clearer that the 2 is because our scale is 2x Goodreads's?

Comment (labradav) [reply src/main/java/com/karankumar/bookproject/backend/service/ImportService.java:140]: Sure. I'll do that.

**Comment by knjk04** [comment]
@labradav Thanks, we'll review this at some point this week

**Comment by knjk04** [comment]
@labradav Is this working for you with the sample file attached in the issue description?

**Comment by labradav** [comment]
> @labradav Is this working for you with the sample file attached in the issue description?

@knjk04 Actually currently it won't work for that file as for the bookshelves column, there is no separator present in the values and in the implementation, I have assumed that the shelf names should be separated by "," character. This is required to parse the predefined book shelf and custom book shelf from all the possible shelf names.

**Comment by knjk04** [comment]
@labradav Could you please get it to work with the sample file? This is the format that csv file exports from Goodreads will be in. I think you would need to read in all of bookshelves separated by spaces. If it was separated by a comma, it would be a different value (e.g. 'My Review')

**Comment by labradav** [comment]
> @labradav Could you please get it to work with the sample file? This is the format that csv file exports from Goodreads will be in. I think you would need to read in all of bookshelves separated by spaces. If it was separated by a comma, it would be a different value (e.g. 'My Review')

@knjk04 I understand. But if the names contains "To read" or "Did not finish", these shelf names would be read "to", "read", "did" and so on. So essentially, the books will only be added if they contain either "Reading" or "Read". Is that the required behaviour?

**Comment by knjk04** [comment]
@labradav Oh, I see what you mean. 

> Is that the required behaviour?

No, that's not.

Here's [another exported file](https://gist.github.com/knjk04/d893646560e329925a943480e113cb23) with the 'To read' bookshelf. Note that they're using single dashes ('-') for 'to-read' and they have 'currently-reading', which is our 'Reading').

Could you please get it to work with both this exported file and the one in the issue description?

**Comment by labradav** [comment]
> Could you please get it to work with both this exported file and the one in the issue description?

Sure I'll do that. What about the "Did not finish" case though? Can you share a sample for that as well?

**Comment by knjk04** [comment]
Goodreads doesn't currently have that, so we can either check for that as three separate words or ignore this for now and we create a separate issue for that

**Comment by labradav** [comment]
> Goodreads doesn't currently have that, so we can either check for that as three separate words or ignore this for now and we create a separate issue for that

Got it. Thanks!

**Comment by labradav** [comment]
@knjk04 The format of 'Date Read' column in the file in the issue description has the format: 'YYYY-MM-DD' while the file you have attached in the previous comment has 'YYYY/MM/DD'. Which one it would be or can it be both?

**Comment by labradav** [comment]
> If a date read has been speciifed, the book should go in the 'to read's shelf

In the above case, shouldn't the book be assigned to the 'read' shelf rather than the 'to read' shelf as it has already been read?

**Comment by labradav** [comment]
Hey @knjk04 ! Did you get a chance to look at my last 2 questions?

**Comment by knjk04** [comment]
Hey @labradav, apologies for the delay. I had a look at your questions, but my weekend was quite hectic, so I didn't get a chance to respond.

> @knjk04 The format of 'Date Read' column in the file in the issue description has the format: 'YYYY-MM-DD' while the file you have attached in the previous comment has 'YYYY/MM/DD'. Which one would it be or can it be both?

Good spot! This is quite strange because the larger file that I attached in this PR was my exported books from Goodreads. Let's go with the latter (with the forward slashes) and we can assume that Goodreads made a mistake/haven't updated their sample file. We can change this later if needed quite simply.

> In the above case, shouldn't the book be assigned to the 'read' shelf rather than the 'to read' shelf as it has already been read?

Another good spot! This was a mistake on my part. You're right that it should go in the 'read' shelf. I've updated the issue description accordingly

**Comment by labradav** [comment]
> Hey @labradav, apologies for the delay. I had a look at your questions, but my weekend was quite hectic, so I didn't get a chance to respond.
> 
> > @knjk04 The format of 'Date Read' column in the file in the issue description has the format: 'YYYY-MM-DD' while the file you have attached in the previous comment has 'YYYY/MM/DD'. Which one would it be or can it be both?
> 
> Good spot! This is quite strange because the larger file that I attached in this PR was my exported books from Goodreads. Let's go with the latter (with the forward slashes) and we can assume that Goodreads made a mistake/haven't updated their sample file. We can change this later if needed quite simply.
> 
> > In the above case, shouldn't the book be assigned to the 'read' shelf rather than the 'to read' shelf as it has already been read?
> 
> Another good spot! This was a mistake on my part. You're right that it should go in the 'read' shelf. I've updated the issue description accordingly

That's completely alright. :)

**Comment by labradav** [comment]
> Let's go with the latter (with the forward slashes)

I have already made the required changes for this.

**Comment by labradav** [comment]
@knjk04 I have made the mentioned changes and the file is now getting uploaded successfully. I'm marking the PR as `Ready for review`.

File: src/test/java/com/karankumar/bookproject/backend/util/CsvUtilsTest.java
Comment (knjk04) [inline src/test/java/com/karankumar/bookproject/backend/util/CsvUtilsTest.java:35]: Could you please rename this test method so that it follows on as a sentence from "CsvUtils should" (currently, we have "CsvUtils should" ... "testReadWhenInputStreamIsCsv"?

**Review by knjk04 — COMMENTED** [review]
Thanks for making the changes. I've also committed, so I recommend pulling before working on this again. Does this work for you when you try to import the large sample file?

**Comment by labradav** [comment]
> Does this work for you when you try to import the large sample file?

Yes it's working for the larger file. Is it not working for you?

File: pom.xml
Comment (labradav) [reply pom.xml]: I have bumped it to 2.12.0, but also added `jackson-core` to support the upgrade.

File: src/test/java/com/karankumar/bookproject/backend/util/CsvUtilsTest.java
Comment (labradav) [reply src/test/java/com/karankumar/bookproject/backend/util/CsvUtilsTest.java]: Sure. I have renamed it to `readSuccessfullyWhenInputStreamIsCsv`.

**Comment by knjk04** [comment]
It doesn't work for me, I'm afraid. Could you please share a gif where you show uploading the large sample file?

**Comment by labradav** [comment]
> It doesn't work for me, I'm afraid. Could you please share a gif where you show uploading the large sample file?

https://drive.google.com/file/d/1qvx0DGarmoTV9hP0Lx0CVDXdMdktiIx2/view?usp=sharing
https://drive.google.com/file/d/17xVH9X9yTsxnxtRQHKTViUWvGVvIq6Bs/view?usp=sharing

The first link shows the file getting uploaded successfully and the second one shows the books visible after upload. Also, can you share the stack trace when you're trying to upload. Maybe I'm missing something.

**Comment by knjk04** [comment]
Thanks for the videos.

Nothing changes in my run window after uploading a file. I'll send a video shortly. Just to double-check -- have you pushed your latest changes to remote?

**Comment by labradav** [comment]
> Just to double-check -- have you pushed your latest changes to remote?

Umm..all the latest changes are on the forked branch as of now, i.e. `labradav:import-goodreads-77`.

**Comment by knjk04** [comment]
Yes

Here's a video: https://drive.google.com/file/d/1lpbqVvMcp9MgnmoUFYXiArJpI0uT9NIR/view?usp=sharing

**Comment by labradav** [comment]
Kindly grant the access as well. My email id is: rahulyadav7477@gmail.com

**Comment by knjk04** [comment]
Oops, done

**Comment by labradav** [comment]
Its quite weird actually. When you are selecting the file, even the vaadin upload button does not show as if its processing/uploading the file. I tried again on my system, and this time I ran the application using the Intellij button instead and it worked fine for me. Are you sure you have pulled the latest changes from the remote branch? I don't understand why it isn't doing anything at all for you.

**Comment by knjk04** [comment]
@amit1307 Could you please try this on macOS?

**Comment by amit1307** [comment]
> @amit1307 Could you please try this on macOS?

sure, could you please share the file to uploaded?

**Comment by knjk04** [comment]
Sure
https://gist.github.com/knjk04/d893646560e329925a943480e113cb23

**Comment by amit1307** [comment]
> Sure
> https://gist.github.com/knjk04/d893646560e329925a943480e113cb23

seems to work for me, I can see was able to upload successfully and can now see tons of books under `All Books`

**Review by knjk04 — APPROVED** [review]
Thanks for working on this! We'll merge this in once you've resolved the merge conflicts

**Comment by labradav** [comment]
Sure. I have resolved the merge conflicts.

**Review by knjk04 — APPROVED** [review]
Thank you very much for working on this!

**Comment by labradav** [comment]
> Thank you very much for working on this!

Thanks @knjk04 . It was a great experience for me as well. Looking forward to some more contributions :)

**Comment by knjk04** [comment]
@labradav That's great to hear that you found it to be a great experience!

If you're interested, we have a couple more import-related issues (currently blocked, but we can unblock them):
- [Import from exported Book Project file](https://github.com/Project-Books/book-project/issues/345)
- [Import from exported LibraryThing file](https://github.com/Project-Books/book-project/issues/348)

**Comment by labradav** [comment]
> @labradav That's great to hear that you found it to be a great experience!
> 
> If you're interested, we have a couple more import-related issues (currently blocked, but we can unblock them):
> 
> * [Import from exported Book Project file](https://github.com/Project-Books/book-project/issues/345)
> * [Import from exported LibraryThing file](https://github.com/Project-Books/book-project/issues/348)

Sure. I'll look into them and select either one soon.

**Pull request merged**
